test/embedding/embedtest.cc

```diff
@@ -73,21 +98,37 @@ int RunNodeInstance(MultiIsolatePlatform* platform,
     HandleScope handle_scope(isolate);
     Context::Scope context_scope(setup->context());
 
-    MaybeLocal<Value> loadenv_ret = node::LoadEnvironment(
-        env,
-        "const publicRequire ="
-        "  require('module').createRequire(process.cwd() + '/');"
-        "globalThis.require = publicRequire;"
-        "globalThis.embedVars = { nön_ascıı: '🏳️‍🌈' };"
-        "require('vm').runInThisContext(process.argv[1]);");
+    MaybeLocal<Value> loadenv_ret;
+    if (snapshot) {
+      loadenv_ret = node::LoadEnvironment(env, node::StartExecutionCallback{});
+    } else {
+      loadenv_ret = node::LoadEnvironment(
+          env,
+          "const publicRequire ="
+          "  require('module').createRequire(process.cwd() + '/');"
+          "globalThis.require = publicRequire;"
+          "globalThis.embedVars = { nön_ascıı: '🏳️‍🌈' };"
+          "require('vm').runInThisContext(process.argv[1]);");
```

Comment:
> because compiling the main script in the embedder test uses the code path that compiles it as a builtin

Is it about this code path? I think in this case it probably makes sense to replace the `node::LoadEnvironment()` call here with `CreateSnapshot()`, or make it so that the `StartExecution` + event loop spinning + blob creation is done is done in `LoadEnvironment()` while `CreateSnapshot()` only serializes that blob? Then the code paths in `StartsExecution()` in this case can be modified to execute `internal/main/mksnapshot` instead of `internal/main/environment`. We could modify `mksnapshot.js` so that it knows when building a snapshot for the embedder, it can just get the source somewhere else (maybe from the `mksnapshot` binding) instead of trying to read from disk: https://github.com/nodejs/node/blob/19623e386fb78b59ba6b79de7e3fd00257d812fc/lib/internal/main/mksnapshot.js#L128 I think the rest of the `mksnapshot.js` should work just fine for the embedder.

Reply:
> Then the code paths in `StartExecution()` in this case can be modified to execute `internal/main/mksnapshot` instead of `internal/main/environment`. We could modify `mksnapshot.js` so that it knows when building a snapshot for the embedder, it can just get the source somewhere else (maybe from the `mksnapshot` binding) instead of trying to read from disk:

Yeah, my question above is essentially “should we do :arrow_up: ” :smile: Sounds good, will implement that first thing next week!

Reply:
@joyeecheung Done – PTAL! :)